src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -393,7 +474,361 @@ public virtual bool TryInsertExpansion(int startPositionInSubjectBuffer, int end
+
+            if (_state._methods.IsDefaultOrEmpty)
+            {
+                // Signature Help is showing a set of overloads that don't match the overloads from the point where the
+                // argument completion session first started. It's unclear how this state should be handled, so we stop
+                // processing Signature Help updates for the current session.
+                // TODO: https://github.com/dotnet/roslyn/issues/50636
+                _state.ClearSymbolInformation();
+                return;
+            }
+
+            var textViewModel = TextView.TextViewModel;
+            if (textViewModel == null)
+            {
+                Debug.Assert(TextView.IsClosed);
+                return;
+            }
+
+            var buffer = EditorAdaptersFactoryService.GetBufferAdapter(textViewModel.DataBuffer);
+            if (buffer is not IVsExpansion expansion)
+            {
+                return;
+            }
+
+            // Track current argument values so input created/updated by a user is not lost when cycling through
+            // Signature Help overloads:
+            //
+            // 1. For each parameter of the method currently presented as a snippet, the value of the argument as
+            //    it appears in code.
+            // 2. Place the argument values in a map from parameter name to current value.
+            // 3. (Later) the values in the map can be read to avoid providing new values for equivalent parameters.
+            if (_state._method is not null)
+            {
+                foreach (var previousParameter in _state._method.Parameters)
+                {
+                    if (ExpansionSession.GetFieldValue(previousParameter.Name, out var previousValue) == VSConstants.S_OK)
+                    {
+                        _state._arguments = _state._arguments.SetItem(previousParameter.Name, previousValue);
+                    }
+                }
+            }
+
+            var textSpan = new VsTextSpan[1];
+            if (ExpansionSession is null || ExpansionSession.GetSnippetSpan(textSpan) != VSConstants.S_OK)
+            {
+                return;
+            }
+
+            var adjustedTextSpan = textSpan[0];
+            var firstField = _state._method?.Parameters.FirstOrDefault()?.Name ?? "placeholder";
+            if (ExpansionSession.GetFieldSpan(firstField, textSpan) != VSConstants.S_OK)
+            {
+                return;
+            }
+
+            adjustedTextSpan.iStartLine = textSpan[0].iStartLine;
+            adjustedTextSpan.iStartIndex = textSpan[0].iStartIndex;
+
+            var lastField = _state._method?.Parameters.LastOrDefault()?.Name ?? "placeholder";
+            if (ExpansionSession.GetFieldSpan(lastField, textSpan) != VSConstants.S_OK)
+            {
+                return;
+            }
+
+            adjustedTextSpan.iEndLine = textSpan[0].iEndLine;
+            adjustedTextSpan.iEndIndex = textSpan[0].iEndIndex;
```

Comment:
It feels like this really needs a comment here somewhere to explain the intent of this. It's...not obvious.

Reply:
Also making a constant out of "placeholder" ("PlaceholderForEmptyArguments"?) would make this a bit less magic.

Reply:
⏱ will document

Reply:
Is there a PR I can make those comments on?